stake-pool/single-pool/src/processor.rs

```diff
@@ -33,42 +33,32 @@ use {
     spl_token_2022::{extension::StateWithExtensions, state::Mint},
 };
 
-// XXX TODO FIXME is this correct? i copied it from the existing stake pool
-// it uses the stake account lamport balance, which includes rent. which might not make sense?
-// since the account can never be dealloced. but it effectively doesnt use the lamports for the first mint
-// eg, if rent is 100, and i deposit 500, i hit the first branch and get 500 tokens
-// then if jon deposits 500, he hits the second, and gets 500 * 500 / 600 = 416.66
-// then if i deposit another 500 i get... 500 * 916.66 / 1100 = 416.66
-// so the calculation is good except at the edge case
-// but... if i did the thing where i took 1 lamp deposit and minted 1 token then
-// the next deposit would massively overweight the rent! i deposit 500 and get 500 * 1 / 101 = 4.95!
-// i think the solution is to calculate off of stake amount rather than lmaports
-/// Calculate pool tokens to mint, given the outstanding tokens, stake account lamports, and deposit amount
+/// Calculate pool tokens to mint, given outstanding token supply, pool active stake, and deposit active stake
 fn calculate_deposit_amount(
-    token_supply: u64,
-    validator_lamports: u64,
-    deposit_lamports: u64,
+    pre_token_supply: u64,
+    pre_pool_stake: u64,
+    user_stake_to_deposit: u64,
 ) -> Option<u64> {
-    if validator_lamports == 0 || token_supply == 0 {
-        Some(deposit_lamports)
+    if pre_pool_stake == 0 || pre_token_supply == 0 {
+        Some(user_stake_to_deposit)
     } else {
         u64::try_from(
-            (deposit_lamports as u128)
-                .checked_mul(token_supply as u128)?
-                .checked_div(validator_lamports as u128)?,
+            (user_stake_to_deposit as u128)
+                .checked_mul(pre_token_supply as u128)?
+                .checked_div(pre_pool_stake as u128)?,
         )
         .ok()
     }
 }
```

Comment:
on your next pass, please pay special attention to this and `calculate_withdraw_amount` because i have changed how they work from the multi-pool version

the comment illustrates a couple weird cases where token calculations break down: when no tokens have yet been minted, and when the amount of rent-exemption lamports in the stake account overwhelms the amount of stake in it

i think for this program it actually makes sense to do all calculations based on stake, rather than lamports, because:
* in single-pool, user rent returns to the user, whereas in multi-pool, user rent moves to the reserve. thus deposited stake and deposited lamports should be identical
* in single-pool, the program stake account can never be closed, so the rent deposit tendered at pool initialization is effectively forfeit, whereas in multi-pool, rent may be reclaimed with validators are wound down, so that value is still "live" for the pool